### DISCUSSION TITLE
Support global config in /etc/

### DISCUSSION THREAD

**Discussion by linsui** [helix-editor/helix#7010] [General]
Looks like helix currenly only reads config from home. Can it also read config from `/etc/`?

**Comment by mmaslyukov** [comment]
Helix is a cross-platform editor, and `/etc/` doesn't make sense for Windows. As a workaround, you might have an `alias hx="hx -c /etc/<path/to/config.toml>"`

**Reply by LunarEclipse363** [reply]
The Windows equivalent would roughly be the registry (or maybe group policy).
There is no direct equivalent as system administration on Windows is completely different conceptually to Linux, mainly Linux takes an "everything is a file" approach while Windows does not.

**Comment by pascalkuthe** [comment]
Putting config files (for user programs) under global paths like `/etc` is explicitly discouraged and a bad practice. If you still want to do that then the workaround suggested by @mmaslyukov is the best way to achieve that. Alternatively, you could set `$XDG_CONFIG_DIRS`

**Reply by LunarEclipse363** [reply]
Why do you think of it as a bad practice?

I understand that it would be bad to use an `/etc` config for a user program in general, or as the primary configuration for a single-user system, since `/etc` is normally reserved for system-wide configuration.

However configs in `/etc` for user programs generally have lower priority than configs in the user's directory and are often used as an administrator-provided default/fallback. I don't think this usage of them should be discouraged, it can be very helpful!

My use case for this specifically is that I have servers where each service gets its own user account (and runs Podman rootless containers, thus I need to log in interactively for management), and I would really like to just be able to place a config file in one place and have it used as the default for every user. Helix has remarkably good defaults, however the default color theme tends to be somewhat difficult to read on light theme terminals, so I find myself always needing to change it before I can use it.

Allowing a config in `/etc` as a fallback to be used only if no config is found in the user directory would solve all those problems and be very simple implementation wise, it shouldn't burden you much when it comes to maintaining Helix long term. I could even write a patch if you're willing to accept one, as this feature would save me a bunch of time.

**Comment by linsui** [comment]
Thanks for your answer. Does the workaround also works for theme and languages?

**Reply by CptPotato** [reply]
These are located in the runtime directory. [Check the manual for details](https://docs.helix-editor.com/master/install.html#configuring-helixs-runtime-files).

**Reply by linsui** [reply]
I'm confused. In https://docs.helix-editor.com/master/themes.html and https://docs.helix-editor.com/master/languages.html the runtime directory is not mentioned. What's the relationship between ~/.config/helix and the runtime directory? The manual doesn't tell what the runtime directory is so I don't know what it is.

**Reply by CptPotato** [reply]
The config directory contains your user configuration. The runtime directory contains runtime files (language grammars, built in themes, etc.) that are bundled with helix.

If you're unsure where these folders are located on your system run `hx --health` and check the paths at the top of the output.

**Reply by linsui** [reply]
Ah, got it, thanks. So the runtime dir is not used for config files.  Then can use other path for mytheme.toml and languages.toml?

**Comment by Elec3137** [comment]
For setting a global config default in nixos, I was able to do this:
```nix
environment.systemPackages = [
  (pkgs.writeShellScriptBin "hx" ''${pkgs.helix}/bin/hx -c /etc/helix.toml "$@"'')
];
```
https://codeberg.org/electria/nixowos/src/branch/main/cli/helix.nix

However, it looks like setting themes or languages cannot be done this way?

**Reply by Abdillah** [reply]
Do you mean adding themes and languages?

Setting theme as I understood is by using `theme = ...`. Language is the problem, as currently no flag can specify `languages.toml`.

**Reply by Elec3137** [reply]
Yes, that's what I meant. Sorry about that

**Reply by Abdillah** [reply]
No worries.

I'm hopeful that we can set runtime directory and thus `languages.toml` is read from that.

**Comment by habruening** [comment]
I think, there are not so many people who would use the config in /etc. Yes, there are use cases, where it makes sense. But not so many. On a single user system, you wouldn't not use it. And on a multi user system, you don't want to login as root to change the helix configuration for the users, because Helix is an application of the user space. If you want to give people a good initial configuration, /etc/skel is a good place. Because we are not only talking about Helix. If you setup a machine, that is used by many people, you should define a nice and comfortable starting configuration for your users.

**Reply by Elec3137** [reply]
On a single user system, I would use it just because I already share a single NixOS config across devices. [`/etc/skel` is not supported by nixos](https://github.com/NixOS/nixpkgs/issues/33586), and it wouldn't be as nice since I would want to update my helix config from time to time, and I shouldn't need to copy from `/etc/skel` again each time.

On a multi user system, any user should be able to override the system config with their own in `~/.config/helix/config.toml`, so root is unnecessary. This is unlike the behavior of wrapping helix with just the `-c` argument, since that will ignore the default config location entirelyfrom what I understand. I will probably make a bash script to mimick this behavior by using the user config if it exists, otherwise using the system config.

What might be even better than this is if helix would merge the global config with the user config.
```nix
environment.systemPackages = [
  (pkgs.writeShellApplication {
    name = "hx";

    runtimeInputs = [ pkgs.helix ];

    text = ''
      if test -r ~/.config/helix/config.toml; then
        exec -a "$0" hx "$@"
      else
        exec -a "$0" hx -c "${pkgs.writeText "${name}.toml" cfg.extraConfig}" "$@"
      fi
    '';
  })
];
```

**Reply by habruening** [reply]
I did not want to say, that I am against /etc. If a few people would use it, I think it is worth it. I just wanted to illustrate that in practice quite often this is also not what you need.

**Reply by dmig** [reply]
@habruening, you're likely wrong.
Use case: Helix as the default system editor — often run from both root and user accounts. I'd like to set some preferences by default: at least the theme and language servers installed globally.

**Reply by dmig** [reply]
Another thought: something like `HELIX_THEME` env variable would be useful, as it can be passed over ssh connection.

**Reply by Elec3137** [reply]
> Use case: Helix as the default system editor — often run from both root and user accounts.

Note that logging in as root is inadvisable in general, and from the user account you should use `sudoedit` instead.

**Reply by dmig** [reply]
@Elec3137 that's what I'm talking about. What user profile do you think is used in this case?

**Reply by Elec3137** [reply]
> @Elec3137 that's what I'm talking about. What user profile do you think is used in this case?

Of the user that called `sudoedit`; not the root user. That's the purpose of using `sudoedit <file>` instead of `sudo helix <file>`
